test/src/code_field/tabs_to_spaces_test.dart

```diff
@@ -21,203 +20,218 @@ class Foo {
-        text: _codeWithTabs,
-        selection: TextSelection(
-          baseOffset: selectionStart,
-          extentOffset: selectionEnd,
-        ),
-      );
-
-      value = value.tabsToSpaces(_spaceCount);
-
-      expect(value.text, _codeWithDoubleSpaces);
-      expect(value.selection.baseOffset, selectionStart + 1);
-      expect(value.selection.extentOffset, selectionEnd + 5);
+    group('Code without tabs', () {
+      test('without selection', () {
+        TextEditingValue value =
+            const TextEditingValue(text: _codeWithDoubleSpaces);
+
+        value = value.tabsToSpaces(_spaceCount);
+
+        expect(value.text, _codeWithDoubleSpaces);
+        expect(value.selection, const TextSelection.collapsed(offset: -1));
+      });
+
+      test('with cursor at the start', () {
+        TextEditingValue value = const TextEditingValue(
+          text: _codeWithDoubleSpaces,
+          selection: TextSelection.collapsed(offset: 0),
+        );
+
+        value = value.tabsToSpaces(_spaceCount);
+
+        expect(value.text, _codeWithDoubleSpaces);
+        expect(value.selection, const TextSelection.collapsed(offset: 0));
+      });
+
+      test('with cursor at the middle', () {
+        const cursorPosition = 5;
+        TextEditingValue value = const TextEditingValue(
+          text: _codeWithDoubleSpaces,
+          selection: TextSelection.collapsed(
+            offset: cursorPosition,
+            affinity: TextAffinity.upstream,
+          ),
+        );
+
+        value = value.tabsToSpaces(_spaceCount);
+
+        expect(value.text, _codeWithDoubleSpaces);
+        expect(
+          value.selection,
+          const TextSelection.collapsed(
+            offset: cursorPosition,
+            affinity: TextAffinity.upstream,
+          ),
+        );
+      });
+
+      test('with cursor at the end', () {
+        const cursorPosition = _codeWithDoubleSpaces.length;
+        TextEditingValue value = TextEditingValue(
+          text: _codeWithDoubleSpaces,
+          selection: const TextSelection.collapsed(offset: cursorPosition)
+              .copyWith(isDirectional: true),
+        );
+
+        value = value.tabsToSpaces(_spaceCount);
+
+        expect(value.text, _codeWithDoubleSpaces);
+        expect(
+          value.selection,
+          const TextSelection.collapsed(offset: cursorPosition)
+              .copyWith(isDirectional: true),
+        );
+      });
+
+      test('with non-empty normalized selection', () {
+        const selectionStart = 5;
+        const selectionEnd = 10;
+        TextEditingValue value = const TextEditingValue(
+          text: _codeWithDoubleSpaces,
+          selection: TextSelection(
+            baseOffset: selectionStart,
+            extentOffset: selectionEnd,
+          ),
+        );
+
+        value = value.tabsToSpaces(_spaceCount);
+
+        expect(value.text, _codeWithDoubleSpaces);
+        expect(value.selection.baseOffset, selectionStart);
```

Comment:
If you construct a selection here, you also implicitly test that `affinity` and `isDirectional` have not changed which is beneficial. This is why we should avoid testing individual properties of standard classes that can be compared as a whole.
Please check each test of individual properties if it can be changed to comparison to a `TextSelection` object.